general/bundler_imports.md

```diff
@@ -0,0 +1,160 @@
+
+### A real world example
+
+To see how npm bundler imports, we'll look at a real case in `DXP` source code.
+
+This question was asked in one of our Slack channels:
+
+> I was trying to debug why using `@claui/charts` was getting an old version of `billboard.js` instead of the one specified in the `package.json`.
+> I noticed if you build `frontend-taglib-clay` and inspect the build folder dependencies. Taking a look at `.../build/node/packageRunBuild/resources/node_modules/@frontend-taglib-clay$clayui%2Fcharts@3.1.2/lib/BillboardWrapper.js`, it shows that `billboard.js` is being imported from `frontend-taglib-chart$billboard.js`. This seems wrong for the bundler to grab it from that module... I would expect it to grab the `billboard.js` version specified by `@clayui/charts`.
+
+And this was the reply:
+
+In your case, your `frontend-taglib-clay` project is importing `billboard.js` from `frontend-taglib-chart` (as per [this configuration](https://github.com/liferay/liferay-portal/blob/af3d78b4b37069b728762df550406459b2f2dd2c/modules/apps/frontend-taglib/frontend-taglib-clay/.npmbundlerrc#L24)), so it is expected that `@frontend-taglib-clay$clayui%2Fcharts@3.1.2`, which is inside `frontend-taglib-clay` (your project; the project declaring the import) grabs `billboard.js` from `frontend-taglib-chart` (given that it's what you are asking for...).
+
+If you don't want to get it from there and use Clay's copy of `billboard.js` simply remove that import. The rationale behind this is that imports are intended to fully divert a package from your project to anywhere else, as opposed to sometimes diverting, sometimes not, which would require a much more complex configuration.
+
+Said that, there are multiple combinations you can do when importing (for example, you could import `billboard.js` only when it appears in your project's source code, but not in your project's dependencies in `node_modules`), but as I said, they would require more complex configuration and we haven't implemented that, because the only use case we need is the current one.
+
+You could argue that you are showing that we need another use case, but I think that may be wrong, because we want to have exactly one `billboard.js` for the whole `DXP`. So, even if in your case what you propose is more correct, we don't want it (I think). AFAIK, we always want to use the `billboard.js` from `frontend-taglib-chart` and if Clay or any other library we are responsible for needs it, it has to make sure it is compatible with the version provided by DXP.
+
+Problem with that last part is that we have no mechanism (apart from our brains) to ensure that Clay, DXP, or any other library we may have are always aligned. But I think that we need to fix this issue in an upper level, not at the bundler's level because, as I said, if we do it in the bundler, we will be technically correct, but we will end up deploying `DXP` with two or more `billboard.js`s.
```

Comment:
You could probably edit this for brevity, since it repeats what you've explained in previous sections.

Reply:
I noticed the duplication but since it refers to a real case in the portal I didn't want to summarize it because there's a risk that the reader gets lost (even if it has been explained in the previous section).

My assumption is that we see the redundancy because we already know how imports work and have an (almost) intuitive mental model of it, but people who hasn't dealt with them will probably need to read the doc more than once, so repeating all the concepts again, in a different scenario, may help to understand not totally understood concepts that have been read before...

Reply:
Maybe someone else not used to configuring this feature may tell us if my assumption is correct or wrong...

They can even say they don't understand anything after reading it several times :sweat_smile: .